src/cocotb/_scheduler.py

```diff
@@ -341,170 +339,105 @@ def _test_completed(self, trigger=None):
+            )
 
-                # This trigger isn't needed any more
-                trigger._unprime()
+        # queue all tasks to wake up
+        for task in scheduling:
+            # unset trigger
+            task._trigger = None
+            self._queue(task, outcome=trigger._outcome)
 
-                for task in self._scheduling:
-                    if task._outcome is not None:
-                        # Task was killed by another task waiting on the same trigger
-                        continue
-                    if _debug:
-                        self.log.debug(f"Scheduling task {task}")
-                    self._schedule(task, trigger=trigger)
-                    if _debug:
-                        self.log.debug(f"Scheduled task {task}")
+        # This trigger isn't needed any more
+        trigger._unprime()
 
-                    # remove our reference to the objects at the end of each loop,
-                    # to try and avoid them being destroyed at a weird time (as
-                    # happened in gh-957)
-                    del task
+    def _event_loop(self) -> None:
+        """Run the main event loop
 
-                self._scheduling = []
+        This should be triggered by only:
+        * The beginning of a test, when there is no trigger to react to
+        * A GPI trigger
+        """
 
-                # Handle any newly queued tasks that need to be scheduled
-                while self._pending_tasks:
-                    task = self._pending_tasks.pop(0)
-                    if _debug:
-                        self.log.debug(f"Scheduling queued task {task}")
-                    self._schedule(task)
-                    if _debug:
-                        self.log.debug(f"Scheduled queued task {task}")
+        while self._pending_tasks and not self._terminate:
+            task, outcome = self._pending_tasks.popitem(last=False)
 
-                    del task
+            if _debug:
+                self.log.debug(f"Scheduling task {task}")
+            self._schedule(task, outcome)
+            if _debug:
+                self.log.debug(f"Scheduled task {task}")
 
-                # Schedule may have queued up some events so we'll burn through those
-                while self._pending_events:
-                    if _debug:
-                        self.log.debug(
-                            f"Scheduling pending event {self._pending_events[0]}"
-                        )
-                    self._pending_events.pop(0).set()
+            # remove our reference to the objects at the end of each loop,
+            # to try and avoid them being destroyed at a weird time (as
+            # happened in gh-957)
+            del task
 
-                # remove our reference to the objects at the end of each loop,
-                # to try and avoid them being destroyed at a weird time (as
-                # happened in gh-957)
-                del trigger
```

Comment:
No longer necessary. We could move this up to the last use of the trigger in `_react`, but it would be at the same place as `trigger._unprime` there, which is equivalent.